backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt

```diff
@@ -0,0 +1,191 @@
+    val memReadFn = findMemoryAccessFunction(isRead = true, valueType = memoryValueType)
+    val memRead = builder.irCall(memReadFn).also { memRead ->
+        memRead.dispatchReceiver = builder.irGetObject(symbols.nativeMemUtils)
+        memRead.putValueArgument(0, builder.irCall(symbols.interopInterpretNullablePointed).also {
+            it.putValueArgument(0, nativePointed)
+        })
+    }
+    return castPrimitiveIfNeeded(memRead, returnType)
+}
+
+private fun InteropCallContext.writePrimitiveToMemory(
+        nativePointed: IrExpression,
+        value: IrExpression
+): IrExpression {
+    val memoryValueType = determineInMemoryType(value.type)
+    val memWriteFn = findMemoryAccessFunction(isRead = false, valueType = memoryValueType)
+    val valueToWrite = castPrimitiveIfNeeded(value, memoryValueType)
+    return with(builder) {
+        irCall(memWriteFn).also { memWrite ->
+            memWrite.dispatchReceiver = irGetObject(symbols.nativeMemUtils)
+            memWrite.putValueArgument(0, irCall(symbols.interopInterpretNullablePointed).also {
+                it.putValueArgument(0, nativePointed)
+            })
+            memWrite.putValueArgument(1, valueToWrite)
+        }
+    }
+}
+
+private fun InteropCallContext.determineInMemoryType(type: IrType): IrType {
+    val classifier = type.classOrNull!!
+    return if (classifier in symbols.unsignedIntegerClasses) {
+        symbols.unsignedToSignedOfSameBitWidth.getValue(classifier).owner.defaultType
+    } else {
+        type
+    }
+}
+
+private fun InteropCallContext.castPrimitiveIfNeeded(
+        value: IrExpression,
+        targetType: IrType
+): IrExpression {
+    val valueClass = value.type.classOrNull!!
+    val targetClass = targetType.classOrNull!!
+    return if (valueClass != targetClass) {
+        val conversion = symbols.integerConversions.getValue(valueClass to targetClass)
+        builder.irCall(conversion.owner).apply {
+            if (conversion.owner.dispatchReceiverParameter != null) {
+                dispatchReceiver = value
+            } else {
+                extensionReceiver = value
+            }
+        }
+    } else {
+        value
+    }
+}
+
+private fun InteropCallContext.convertEnumToIntegral(enumValue: IrExpression): IrExpression {
+    val enumClass = enumValue.type.getClass()!!
+    val valueProperty = enumClass.properties.single { it.name.asString() == "value" }
+    return builder.irCall(valueProperty.getter!!).also {
+        it.dispatchReceiver = enumValue
+    }
+}
+
+private fun InteropCallContext.convertIntegralToEnum(value: IrExpression, enumType: IrType): IrExpression {
+    val enumClass = enumType.getClass()!!
+    val companionClass = enumClass.declarations.filterIsInstance<IrClass>().single { it.isCompanion }
+    val byValue = companionClass.simpleFunctions().single { it.name.asString() == "byValue" }
+    val byValueArg = castPrimitiveIfNeeded(value, byValue.valueParameters.first().type)
```

Comment:
Is `castPrimitiveIfNeeded` required here?

Reply:
It's not really required, but it makes `convertIntegralToEnum` a bit more flexible (function's users are not required to perform cast beforehand).